controllers/rollingupgrade_controller.go

```diff
@@ -161,3 +166,18 @@ func (r *RollingUpgradeReconciler) UpdateStatus(rollingUpgrade *v1alpha1.Rolling
 		r.Error(err, "failed to update status", "name", rollingUpgrade.NamespacedName())
 	}
 }
+
+func (r *RollingUpgradeReconciler) IsRollingUpgradeCompleted(rollingUpgrade *v1alpha1.RollingUpgrade) bool {
+	r.Info("checking if rolling upgrade is completed", "name", rollingUpgrade.NamespacedName())
+	if err := r.Cloud.Discover(); err != nil {
```

Comment:
I am not sure if I need to cache here or wait for the cache to refresh by itself.

Reply:
I don't think `Discover()` is needed here since it was already called in `Reconcile()` before `IsRollingUpgradeCompleted()` is called. I think the basic structure of `Reconcile()` is:
1. Check if this CR can be admitted for processing
2. Refresh AWS cache by calling `Cloud.Discover()`.
3. Do the work needed based on the state of resources in the AWS cache.
I think you only would need to call `Cloud.Discover()` once.

Reply:
I think we can remove these last two blocks all together.
The CR should be marked completed within upgrade.go, once the last instance is rotated and we detect no more instances that require rotation.

Then there is no need for `IsRollingUpgradeCompleted` or the above check of:
```
	if r.IsRollingUpgradeCompleted(rollingUpgrade) {
		rollingUpgrade.SetCurrentStatus(v1alpha1.StatusComplete)
	}
```

This is because on the next reconcile we will finish via:
```
// Stop processing upgrades which are in finite state
	currentStatus := rollingUpgrade.CurrentStatus()
	if common.ContainsEqualFold(v1alpha1.FiniteStates, currentStatus) {
		r.AdmissionMap.Delete(req.NamespacedName)
		r.Info("rolling upgrade ended", "name", req.NamespacedName, "status", currentStatus)
		return reconcile.Result{}, nil
	}
```

Definitely avoid calling `Discover()` more than once, when you need to refresh your state, requeue